src/uct/base/uct_iov.inl

```diff
@@ -111,26 +111,39 @@ size_t uct_iov_to_iovec(struct iovec *io_vec, size_t *io_vec_cnt_p,
  * @param [in] iov_cnt     A number of elements in the array.
  * @param [in] iov_offset  An offset in the array.
  * @param [in] buf         A buffer to copy data to.
+ * @param [in] copy_limit  A maximum amount of data that should be copied.
+ *
+ * @return The amount of copied bytes.
  */
 static UCS_F_ALWAYS_INLINE
-void uct_iov_to_buffer(const uct_iov_t *iov, size_t iovcnt, size_t iov_offset,
-                       void *buf)
+size_t uct_iov_to_buffer(const uct_iov_t *iov, size_t iovcnt, size_t iov_offset,
+                         void *buf, size_t copy_limit)
 {
     size_t iov_it = 0;
+    size_t copied = 0;
 
     while (iov_offset > iov[iov_it].length) {
         iov_offset -= iov[iov_it].length;
         ++iov_it;
     }
 
-    memcpy(buf, UCS_PTR_BYTE_OFFSET(iov[iov_it].buffer, iov_offset),
-           iov[iov_it].length - iov_offset);
-    buf = UCS_PTR_BYTE_OFFSET(buf, iov[iov_it].length - iov_offset);
+    copied = ucs_min(copy_limit, iov[iov_it].length - iov_offset);
+    memcpy(buf, UCS_PTR_BYTE_OFFSET(iov[iov_it].buffer, iov_offset), copied);
+
+    for (++iov_it; iov_it < iovcnt && copied + iov[iov_it].length <= copy_limit;
+         ++iov_it) {
+        memcpy(UCS_PTR_BYTE_OFFSET(buf, copied), iov[iov_it].buffer,
+               iov[iov_it].length);
+        copied += iov[iov_it].length;
+    }
 
-    for (++iov_it; iov_it < iovcnt; ++iov_it) {
-        memcpy(buf, iov[iov_it].buffer, iov[iov_it].length);
-        buf = UCS_PTR_BYTE_OFFSET(buf, iov[iov_it].length);
+    if (iov_it < iovcnt) {
+        memcpy(UCS_PTR_BYTE_OFFSET(buf, copied), iov[iov_it].buffer,
+               copy_limit - copied);
+        copied = copy_limit;
```

Comment:
can we unite the 3 memcpy places to one loop?
```
for (; iov_it < iovcnt && !last; ++iov_it) {
    length = iov[iov_it].length - iov_offset;
    if (copied + length > copy_limit) {
        length = copy_limit - copied;
        last = 1
    }
    memcpy(UCS_PTR_BYTE_OFFSET(buf, copied), iov[iov_it].buffer + iov_offset, length);
    copied += length;
    iov_offset = 0;
}
```

Reply:
yes, sure.
I tried to get rid of any additional checks inside the loop body and to use iov_offset only once. Just to help compiler to optimize the code.

Reply:
yes, i noticed that :) but maybe it's possible to both allow compiler optimize and also unite loops